src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,214 @@
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for objects returned from function calls in a match scrutinee that have the attribute
+    /// `has_significant_drop`.
+    ///
+    /// ### Why is this bad?
+    /// The attribute `has_significant_drop` is added to types whose Drop impls have an important
+    /// side-effect, such as unlocking a mutex, making it important for users to be able to
+    /// accurately understand their lifetimes. When a temporary is returned in a function call in a
+    /// match scrutinee, its lifetime lasts until the end of the match block, which may be
+    /// surprising. It can also lead to deadlocks if the match scrutinee uses a function call that
+    /// returns a MutexGuard and then tries to lock again in one of the match arms, because the
+    /// MutexGuard will not be dropped until the end of the match block and thus will not unlock.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     match mutex.lock().unwrap().foo() {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar(); // Deadlock!
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let Some(expr_span) = has_significant_drop_in_scrutinee(cx, expr) {
+            span_lint_and_help(
+                cx,
+                SIGNIFICANT_DROP_IN_SCRUTINEE,
+                expr_span,
+                "The lifetimes of temporaries in a match scrutinee last to the end of the \
```

Comment:
The warning message is not meant to explain the whole issue. Just point out in a one-liner (optimally) what's wrong with the code. For more details there is the lint documentation.

A good rule of thumb for good error messages is the rule on how error messages should look like:
- start lower case
- don't have any punctuation (at the end)

In this case a good message would be something along the lines of: 

```
"temporary with significant drop glue in match scrutinee"
```

And then in a `note` you can add `"the lifetime of this temporary gets extended to the end of the match expression"`.

Feel free to improve the wording.

Take a look at `span_lint_and_then` in the rest of Clippy on how to contruct more complex error messages and suggestions.

Reply:
Thank you for the detailed explanation!